### DISCUSSION TITLE
Returing ndarray

### DISCUSSION THREAD

**Discussion by thetwom** [wjakob/nanobind#629] [Q&A] [answered]
This question might be too simple, but I read the docs several times and can't find a clear answer.

Assume I want to return a ndarray, is it valid to do something like:
```
...
.def("test", []() {
   std::array a {10.0, 20.0};
   return nb::ndarray<nb::numpy, double>(a.data(), { a.size() }, nb::handle());
})
...
```

Will I get ownership problems or will nanobind copy the data and everything is fine? I understand that for larger objects as vectors, one would want to avoid copies, but for objects like arrays or fixed sized vectors, the code would stay much cleaner.

Besides I wonder, if one could avoid the ownership code (e.g. capsules) for large structures, by letting the ndarray directly allocate it's own memory and then just e.g. use std::span and fill it, something like that:
```
...
.def("test", []() {
   nb::ndarray<nb::numpy, nb::shape<100>, double> result; // array with allocated memory
   std::span<double> view(result.data(), result.data()+result.size(); 
   std::fill(view.begin(), view.end(), 100); // arbitrary function to fill the ndarray ....
   return result;
}
```

**Comment by wjakob** [comment, marked as answer]
Regarding the first problem: yes, you will get ownership problems. The `std::array` lives on the stack and will expire. You need some kind of container with a Python owner, as is explained in the documentation.

What you request below is -- sort of -- possible. The problem is that `nb::ndarray` is just metadata describing where an array is to be found. You need to convert it into a Python object using a framework of choice (eg. NumPy) for an actual copy to be made. Something like the following (untested)

```
nb::ndarray<nb::numpy> result = nb::ndarray(...);
nb::object result = nb::cast(result, nb::rv_policy::copy);
return result;
```

**Reply by thetwom** [reply]
Thanks a lot for the explanation. This helps a lot!
